agent/consul/rate/handler.go

```diff
@@ -0,0 +1,103 @@
+// package rate implements server-side RPC rate limiting.
+package rate
+
+import (
+	"context"
+	"net"
+	"sync/atomic"
+
+	"github.com/hashicorp/consul/agent/consul/rate/multilimiter"
+)
+
+// Mode determines the action that will be taken when a rate limit has been
+// exhausted (e.g. log and allow, or reject).
+type Mode int
+
+const (
+	// ModePermissive causes the handler to log the rate-limited operation but
+	// still allow it to proceed.
+	ModePermissive Mode = iota
+
+	// ModeEnforce causes the handler to reject the rate-limted operation.
+	ModeEnforce
+)
+
+// OperationType is the type of operation the client is attempting to perform.
+type OperationType int
+
+const (
+	// OperationTypeRead represents a read operation.
+	OperationTypeRead OperationType = iota
+
+	// OperationTypeWrite represents a write operation.
+	OperationTypeWrite
+)
+
+// Operation the client is attempting to perform.
+type Operation struct {
```

Comment:
We might also want to add a `Category` field to this struct (e.g. `CategoryKV`, `CategoryCatalog`).